hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenClaimAirdropHandler.java

```diff
@@ -36,25 +71,169 @@
  * HederaFunctionality#TOKEN_CLAIM_AIRDROP}.
  */
 @Singleton
-public class TokenClaimAirdropHandler implements TransactionHandler {
+public class TokenClaimAirdropHandler extends TransferExecutor implements TransactionHandler {
+    private final TokenAirdropValidator validator;
 
     @Inject
-    public TokenClaimAirdropHandler() {
-        // Exists for injection
+    public TokenClaimAirdropHandler(
+            @NonNull final TokenAirdropValidator validator,
+            @NonNull final CryptoTransferValidator cryptoTransferValidator) {
+        super(cryptoTransferValidator);
+        this.validator = validator;
     }
 
     @Override
-    public void preHandle(@NonNull PreHandleContext context) throws PreCheckException {}
+    public void preHandle(@NonNull PreHandleContext context) throws PreCheckException {
+        requireNonNull(context);
+        final var op = context.body().tokenClaimAirdrop();
+        requireNonNull(op);
+        final var pendingAirdrops = op.pendingAirdrops();
+
+        final var accountStore = context.createStore(ReadableAccountStore.class);
+
+        for (final var pendingAirdrop : pendingAirdrops) {
+            AccountID receiverId = pendingAirdrop.receiverIdOrThrow();
+            Account account = accountStore.getAccountById(receiverId);
+            requireNonNull(account);
+            Key key = account.key();
+            validateTruePreCheck(key != null, KEY_NOT_PROVIDED);
+            context.requireKey(key);
+        }
+    }
 
     @Override
-    public void pureChecks(@NonNull TransactionBody txn) throws PreCheckException {}
+    public void pureChecks(@NonNull TransactionBody txn) throws PreCheckException {
+        requireNonNull(txn);
+
+        final var op = txn.tokenClaimAirdrop();
+        requireNonNull(op);
+
+        final List<PendingAirdropId> pendingAirdrops = op.pendingAirdrops();
+        if (pendingAirdrops.isEmpty()) {
+            throw new PreCheckException(EMPTY_PENDING_AIRDROP_ID_LIST);
+        }
+
+        final int numAirdrops = pendingAirdrops.size();
+        final Set<PendingAirdropId> uniqueAirdrops = Set.copyOf(pendingAirdrops);
+        if (numAirdrops != uniqueAirdrops.size()) {
+            throw new PreCheckException(PENDING_NFT_AIRDROP_ALREADY_EXISTS);
```

Comment:
Should be something like `REPEATED_PENDING_AIRDROP`

Reply:
Done